nilearn/_utils/niimg_conversions.py

```diff
@@ -222,42 +248,61 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False):
                     i_error = "Image #" + str(index)
                 raise ValueError("%s is a 4D shape (shape: %s), but this "
                                  "function accepts only 3D images"
-                                % (i_error, this_shape))
-            sizes.append(this_shape[3])
+                                 % (i_error, this_shape))
+            lengths.append(this_shape[3])
 
     # Using fortran order makes concatenation much faster than with C order,
     # because the voxels for a given image are grouped together in memory.
-    data = np.ndarray(first_data_shape[:3] + (sum(sizes), ),
+    data = np.ndarray(target_item_shape + (sum(lengths), ),
                       order="F", dtype=dtype)
-    data[..., :sizes[0]] = _to_4d(first_data)
-    del first_data, first_img
 
-    for index, (iter_niimg, size) in enumerate(zip(niimgs, sizes)):
-        if index == 0:
+    data[..., :lengths[0]] = _to_4d(first_data)
+    cur_4d_index = 0
+    for index, (iter_niimg, size) in enumerate(zip(niimgs, lengths)):
+        # talk to user
+        if (isinstance(iter_niimg, basestring)):
+            nii_str = "image " + iter_niimg
+        else:
+            nii_str = "image #" + str(index)
+        if verbose > 0:
+            print "Concatenating {}/{}: {}".format(index + 1, sum(lengths),
+                                                   nii_str)
+
+        if index == 0:  # we have already loaded the first one
+            cur_4d_index += size
             continue
-        img = check_niimg(iter_niimg)
-        if not np.array_equal(img.get_affine(), affine):
-            if (isinstance(iter_niimg, basestring)):
-                i_error = "image " + iter_niimg
-            else:
-                i_error = "image #" + str(index)
-
-            raise ValueError("Affine of %s is different"
-                             " from reference affine"
-                             "\nReference affine:\n%s\n"
-                             "Wrong affine:\n%s"
-                             % (i_error,
-                             repr(affine), repr(img.get_affine())))
-        this_data = img.get_data()
-        if this_data.shape[:3] != first_data_shape[:3]:
-            if (isinstance(iter_niimg, basestring)):
-                i_error = "image " + iter_niimg
-            else:
-                i_error = "image #" + str(index)
-            raise ValueError("Shape of %s is different from first image shape."
-                             % i_error)
-        data[..., index:index + size] = _to_4d(this_data)
-    return nibabel.Nifti1Image(data, affine)
+
+        # parse current 3D or 4D nifti or path to such
```

Comment:
I don't understand what you mean by "parse" here and also a few times in the lines below. Also I am not sure what useful piece of information this comments bring.